# Lab | Web Scraping

Welcome to the "Books to Scrape" Web Scraping Adventure Lab!

**Objective**

In this lab, we will embark on a mission to unearth valuable insights from the data available on Books to Scrape, an online platform showcasing a wide variety of books. As data analyst, you have been tasked with scraping a specific subset of book data from Books to Scrape to assist publishing companies in understanding the landscape of highly-rated books across different genres. Your insights will help shape future book marketing strategies and publishing decisions.

**Background**

In a world where data has become the new currency, businesses are leveraging big data to make informed decisions that drive success and profitability. The publishing industry, much like others, utilizes data analytics to understand market trends, reader preferences, and the performance of books based on factors such as genre, author, and ratings. Books to Scrape serves as a rich source of such data, offering detailed information about a diverse range of books, making it an ideal platform for extracting insights to aid in informed decision-making within the literary world.

**Task**

Your task is to create a Python script using BeautifulSoup and pandas to scrape Books to Scrape book data, focusing on book ratings and genres. The script should be able to filter books with ratings above a certain threshold and in specific genres. Additionally, the script should structure the scraped data in a tabular format using pandas for further analysis.

**Expected Outcome**

A function named `scrape_books` that takes two parameters: `min_rating` and `max_price`. The function should scrape book data from the "Books to Scrape" website and return a `pandas` DataFrame with the following columns:

**Expected Outcome**

- A function named `scrape_books` that takes two parameters: `min_rating` and `max_price`.
- The function should return a DataFrame with the following columns:
  - **UPC**: The Universal Product Code (UPC) of the book.
  - **Title**: The title of the book.
  - **Price (£)**: The price of the book in pounds.
  - **Rating**: The rating of the book (1-5 stars).
  - **Genre**: The genre of the book.
  - **Availability**: Whether the book is in stock or not.
  - **Description**: A brief description or product description of the book (if available).
  
You will execute this script to scrape data for books with a minimum rating of `4.0 and above` and a maximum price of `£20`. 

Remember to experiment with different ratings and prices to ensure your code is versatile and can handle various searches effectively!

**Resources**

- [Beautiful Soup Documentation](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)
- [Pandas Documentation](https://pandas.pydata.org/pandas-docs/stable/index.html)
- [Books to Scrape](https://books.toscrape.com/)


**Hint**

Your first mission is to familiarize yourself with the **Books to Scrape** website. Navigate to [Books to Scrape](http://books.toscrape.com/) and explore the available books to understand their layout and structure. 

Next, think about how you can set parameters for your data extraction:

- **Minimum Rating**: Focus on books with a rating of 4.0 and above.
- **Maximum Price**: Filter for books priced up to £20.

After reviewing the site, you can construct a plan for scraping relevant data. Pay attention to the details displayed for each book, including the title, price, rating, and availability. This will help you identify the correct HTML elements to target with your scraping script.

Make sure to build your scraping URL and logic based on the patterns you observe in the HTML structure of the book listings!


---

**Best of luck! Immerse yourself in the world of books, and may the data be with you!**

**Important Note**:

In the fast-changing online world, websites often update and change their structures. When you try this lab, the **Books to Scrape** website might differ from what you expect.

If you encounter issues due to these changes, like new rules or obstacles preventing data extraction, don’t worry! Get creative.

You can choose another website that interests you and is suitable for scraping data. Options like Wikipedia, The New York Times, or even library databases are great alternatives. The main goal remains the same: extract useful data and enhance your web scraping skills while exploring a source of information you enjoy. This is your opportunity to practice and adapt to different web environments!

In [4]:
import requests
import pandas as pd
from bs4 import BeautifulSoup

# Website to scrape
url = "https://books.toscrape.com/"

# Download the webpage
response = requests.get(url)


response.encoding = "utf-8"

# Check whether the request was successful
print("Status code:", response.status_code)

# Parse the HTML content
soup = BeautifulSoup(response.text, "html.parser")

# Find all book containers
books = soup.find_all("article", class_="product_pod")

# Create an empty list for the results
book_data = []

# Extract information from each book
for book in books:
    title = book.h3.a["title"]

    price = book.find("p", class_="price_color").text.strip()
    price = float(price.replace("£", ""))

    availability = book.find(
        "p", class_="instock availability"
    ).text.strip()

    rating = book.find("p", class_="star-rating")["class"][1]

    book_data.append({
        "title": title,
        "price_gbp": price,
        "availability": availability,
        "rating": rating
    })

# Convert the results into a pandas DataFrame
books_df = pd.DataFrame(book_data)

# Display the first five rows
books_df.head()

Status code: 200


,title,price_gbp,availability,rating
0,A Light in the Attic,51.77,In stock,Three
1,Tipping the Velvet,53.74,In stock,One
2,Soumission,50.10,In stock,One
3,Sharp Objects,47.82,In stock,Four
4,Sapiens: A Brief History of Humankind,54.23,In stock,Five


In [5]:
from urllib.parse import urljoin

# Convert rating words into numbers
rating_numbers = {
    "One": 1,
    "Two": 2,
    "Three": 3,
    "Four": 4,
    "Five": 5
}


def scrape_books(min_rating, max_price):

    base_url = "https://books.toscrape.com/"
    page_url = base_url + "catalogue/page-1.html"

    book_data = []

    # Continue scraping until there is no next page
    while page_url:

        response = requests.get(page_url)
        response.raise_for_status()
        response.encoding = "utf-8"

        soup = BeautifulSoup(response.text, "html.parser")
        books = soup.find_all("article", class_="product_pod")

        for book in books:

            # Extract and convert the price
            price_text = book.find(
                "p", class_="price_color"
            ).text.strip()

            price = float(
                price_text.replace("£", "").replace("Â", "")
            )

            # Extract and convert the rating
            rating_word = book.find(
                "p", class_="star-rating"
            )["class"][1]

            rating = rating_numbers[rating_word]

            # Ignore books that do not satisfy the filters
            if rating < min_rating or price > max_price:
                continue

            # Create the link to the individual book page
            product_url = urljoin(page_url, book.h3.a["href"])

            product_response = requests.get(product_url)
            product_response.raise_for_status()
            product_response.encoding = "utf-8"

            product_soup = BeautifulSoup(
                product_response.text,
                "html.parser"
            )

            # Extract the product information table
            product_information = {}

            for row in product_soup.select(
                "table.table.table-striped tr"
            ):
                heading = row.th.text.strip()
                value = row.td.text.strip()
                product_information[heading] = value

            # Extract the genre from the breadcrumb
            breadcrumbs = product_soup.select(
                "ul.breadcrumb li"
            )

            genre = breadcrumbs[-2].text.strip()

            # Extract the description
            description_heading = product_soup.find(
                "div",
                id="product_description"
            )

            if description_heading:
                description = (
                    description_heading
                    .find_next_sibling("p")
                    .text.strip()
                )
            else:
                description = "No description available"

            # Add the book to the results
            book_data.append({
                "UPC": product_information.get("UPC"),
                "Title": product_soup.select_one(
                    "div.product_main h1"
                ).text.strip(),
                "Price (£)": price,
                "Rating": rating,
                "Genre": genre,
                "Availability": product_information.get(
                    "Availability"
                ),
                "Description": description
            })

        # Find the next catalogue page
        next_button = soup.select_one("li.next a")

        if next_button:
            page_url = urljoin(
                page_url,
                next_button["href"]
            )
        else:
            page_url = None

    # Set the required column order
    columns = [
        "UPC",
        "Title",
        "Price (£)",
        "Rating",
        "Genre",
        "Availability",
        "Description"
    ]

    return pd.DataFrame(book_data, columns=columns)

In [6]:
filtered_books = scrape_books(
    min_rating=4,
    max_price=20
)

print("Number of matching books:", len(filtered_books))

filtered_books.head()

Number of matching books: 75


,UPC,Title,Price (£),Rating,Genre,Availability,Description
0,ce6396b0f23f6ecc,Set Me Free,17.46,5,Young Adult,In stock (19 available),Aaron Ledbetter’s future had been planned out ...
1,6258a1f6a6dcfe50,The Four Agreements: A Practical Guide to Pers...,17.66,5,Spirituality,In stock (18 available),"In The Four Agreements, don Miguel Ruiz reveal..."
2,6be3beb0793a53e7,Sophie's World,15.94,5,Philosophy,In stock (18 available),A page-turning novel that is also an explorati...
3,657fe5ead67a7767,Untitled Collection: Sabbath Poems 2014,14.27,4,Poetry,In stock (16 available),"More than thirty-five years ago, when the weat..."
4,51653ef291ab7ddc,This One Summer,19.49,4,Sequential Art,In stock (16 available),"Every summer, Rose goes with her mom and dad t..."


In [9]:
print("Minimum rating:", filtered_books["Rating"].min())
print("Maximum price:", filtered_books["Price (£)"].max())
print("DataFrame shape:", filtered_books.shape)

Minimum rating: 4
Maximum price: 19.69
DataFrame shape: (75, 7)


### Conclusion

The `scrape_books` function successfully scraped book information from the Books to Scrape website and returned the results as a pandas DataFrame. After filtering for books with a minimum rating of 4 and a maximum price of £20, 75 books matched the conditions, with ratings ranging from 4 to 5 and a highest price of £19.69.